File: crates/taflow-metrics/src/metrics/common_sense_ratio.rs
use crate::{
    primitives::{ExactOrderStatistics, GainLossState},
    MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy,
};

/// Profit factor multiplied by the exact 95th-to-5th percentile tail ratio.
#[derive(Debug, Clone)]
pub struct CommonSenseRatio {
    input: MetricInputState,
    observations: GainLossState,
    order_statistics: ExactOrderStatistics,
}

impl CommonSenseRatio {
    /// Construct an empty state for decimal simple returns.
    pub fn new(input_kind: MetricInputKind, nan_policy: NanPolicy) -> MetricResult<Self> {
        if input_kind != MetricInputKind::Returns {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason: "common sense ratio requires decimal simple returns",
            });
        }
        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            observations: GainLossState::new(),
            order_statistics: ExactOrderStatistics::new(),
        })
    }

    /// Append one return and refresh the current composite ratio.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(simple_return) = self.input.append(value)? {
            self.observations.append(simple_return);
            self.order_statistics.append(simple_return);
        }
        Ok(self.value())
    }

    /// Append a chronological slice and sort retained values only once.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            if let Some(simple_return) = self.input.append(value)? {
                self.observations.append(simple_return);
                self.order_statistics.append(simple_return);
            }
        }
        Ok(self.value())
    }

    /// Return profit factor multiplied by the absolute tail ratio.
    ///
    /// Empty input, zero gross loss, or a zero-magnitude lower percentile is
    /// undefined. A loss-only sample can validly return zero.
    pub fn value(&mut self) -> Option<f64> {
        let absolute_gross_loss = -self.observations.gross_loss();
        if absolute_gross_loss == 0.0 {
            return None;
        }
        let lower = self
            .order_statistics
            .quantile(0.05)
            .expect("CommonSenseRatio uses a valid fixed lower quantile")?;
        let upper = self
            .order_statistics
            .quantile(0.95)
            .expect("CommonSenseRatio uses a valid fixed upper quantile")?;
        let lower_magnitude = lower.abs();
        if lower_magnitude == 0.0 {
            return None;
        }
        let profit_factor = self.observations.gross_gain() / absolute_gross_loss;
        Some(profit_factor * upper.abs() / lower_magnitude)
    }

    /// Return the current exact scalar without replaying observations.
    pub fn compute(&mut self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while retaining allocated buffers.
    pub fn reset(&mut self) {
        self.input.reset();
        self.observations.reset();
        self.order_statistics.reset();
    }

    /// Return the number of usable simple returns retained.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable return has been retained.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

File: python/taflow/metrics/common_sense_ratio.py
"""Profit-factor and tail-ratio composite return metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import CommonSenseRatio as _Native
from ._input import as_metric_series


class CommonSenseRatio:
    """Compute profit factor multiplied by the 95th-to-5th percentile tail ratio.

    Profit factor is the sum of strictly positive returns divided by the
    absolute sum of strictly negative returns. Tail ratio is the absolute 95th
    linear percentile divided by the absolute 5th linear percentile. The
    independent oracle is QuantStats 0.0.81 ``common_sense_ratio``; correctness
    uses its ``profit_factor`` and ``tail_ratio`` components with preparation
    disabled to avoid QuantStats' price/return inference heuristic.

    Only decimal simple returns are accepted. Empty/all-missing warm-up, a zero
    gross-loss denominator, and a zero-magnitude lower percentile return
    ``None``. A loss-only sample validly returns ``0.0``. This deliberately
    applies TAFlow's general ratio rule to the composite: unlike standalone
    ``ProfitFactor``, a positive-only sample does not expose infinity. NaNs are
    omitted by default or rejected by ``nan_policy="raise"``; infinities and
    returns below -100% are rejected. ``append``, ``extend``, and ``reset`` are
    fluent. Rust retains O(n) observations for exact quantiles, lazily refreshes
    a sorted cache, and sorts once after each bulk extension. Native bulk work
    releases the GIL; Python performs no financial arithmetic.
    """

    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(self, returns: Any, *, column: str | None = None) -> "CommonSenseRatio":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def append(self, value: float) -> "CommonSenseRatio":
        """Append one simple return and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, returns: Any, *, column: str | None = None
    ) -> "CommonSenseRatio":
        """Append chronological simple returns and return this metric."""
        self._state.extend(as_metric_series(returns, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return the current exact composite ratio, or ``None`` if undefined."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current exact scalar without replaying retained returns."""
        return self._state.compute()

    def reset(self) -> "CommonSenseRatio":
        """Clear retained returns, preserve missing policy, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable simple returns retained by Rust."""
        return len(self._state)


__all__ = ["CommonSenseRatio"]
